lib/explorer/data_frame.ex

```diff
@@ -1080,8 +1096,51 @@ defmodule Explorer.DataFrame do
           DataFrame.t()
   def mutate(df, columns) when is_column_pairs(columns) do
     pairs = to_column_pairs(df, columns)
+    mutations = Map.new(pairs)
+    out_df = df_for_mutations(df, mutations)
+
+    Shared.apply_impl(df, :mutate, [out_df, mutations])
+  end
+
+  defp df_for_mutations(df, mutations) do
+    names = df.names
+    mut_names = mutations |> Map.keys() |> MapSet.new()
+
+    new_names =
+      names
+      |> MapSet.new()
+      |> MapSet.union(mut_names)
+      |> MapSet.to_list()
```

Comment:
Using MapSet and Map is problematic here because they will alphabetically reorder the columns. You can add a test where you have a dataframe with keys `[:e, :c, :a]` and you mutate to add the columns `[d: 1, b: 2]`, you should end up with `[:e, :c, :a, :d, :b]`. Note the `Map.new` call in the clause above is also part of the problem.

But I also think this code can be simpler. I think you can do this:

    mut_names = Enum.map(mutations, &elem(&1, 0))
    new_names = Enum.uniq(df.names ++ mut_names)
    mutations = Map.new(mutations)

Something else that I would add is to handle the anonymous function within this function so Explorer.DataFrame does not have to handle it. So overall something like this:

    # First convert mutations to a list and resolve anonymous functions
    mutations =
      for {key, value} <- mutations do
        {key, if(is_function(value), do: value.(df), else: value)}
      end

    # Keep the names ordered
    mut_names = Enum.map(mutations, &elem(&1, 0))

    # Compute the mutable types as you have today (except for functions handled above)
    mut_types =
      for {key, value} <- mutations, into: %{} do
      end

    new_names = Enum.uniq(df.names ++ mut_names)
    new_dtypes = Map.merge(df.dtypes, mut_types)

    out_df = %{df | names: new_names, dtypes: new_dtypes}

    # Send the mutations with functions already resolved
    # and simplify the backend code
    Shared.apply_impl(df, :mutate, [out_df, mutations])

Reply:
Maybe this is a good discussion for https://github.com/elixir-nx/explorer/issues/245

Reply:
I mean: we don't need to decide it here :grimacing:

Reply:
I agree it should be solved with #245. :)

Reply:
Definitely #245. And solving at the DataFrame level would preclude more efficient backend implementations such as groupby + apply in Polars :).